### ISSUE TITLE
[BUG] Calculating maps from constructed labeled globals creates random results.

### ISSUE THREAD

**Issue by mariux** [mineiros-io/terramate#866]
**Describe the bug**

**To Reproduce**

1. Given the following code in a stack

```hcl
stack {
  name        = "providers"
  description = "providers"
  id          = "1e9936ee-5181-4cb4-b0cd-9385b8653226"
}

globals {
  _available_providers = {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.14"
    }
    vault = {
      source  = "hashicorp/vault"
      version = "~> 3.10"
    }
    postgresql = {
      source  = "cyrilgdn/postgresql"
      version = "~> 1.18.0"
    }
    mysql = {
      source  = "petoju/mysql"
      version = "~> 3.0.29"
    }
  }

  required_providers = {for k, v in global._available_providers : k => v if tm_try(global.use_providers[k], false)}
}

globals "use_providers" {
  aws   = true
}

globals "use_providers" {
  mysql = true
}
```

3. Run command `terramate experimental eval global.required_providers`

4. See error of randomly generating content of `required_providers` global.

**Expected behavior**

consitently generate the following code:
```hcl
	required_providers = {
	  aws = {
	    source  = "hashicorp/aws"
	    version = "~> 4.14"
	  }
	  mysql = {
	    source  = "petoju/mysql"
	    version = "~> 3.0.29"
	  }
	}
```

**Environment (please complete the following information):**
 - OS: Linux
 - Terramate Version 0.2.13

**Additional context**
- as detected in #864 
- setting `globals { use_providers = { ... }` will not create random results.

**Comment by mariux** [comment]
I identified a workaround:

```
  required_providers = {
    for k in tm_keys(global.use_providers) :
        k => global._available_providers[k]
        if tm_try(global.use_providers[k], false)
  }
```

triggers postponing the evaluation of `required_providers` as we reference an unindexed `global.use_providers` in the iterator of the `for` loop.

this does generate correctly all of the times.

**Comment by sergeyshevch** [comment]
@geowatson. Can you try this workaround in our state for now?

**Comment by geowatson** [comment]
@mariux this example works fine indeed, thanks!

But for the purpose of simplicity it is better to fix RC (possibly) happening in the base example, because it is supposed to work as intended - generating `required_providers` block only after fully evaluating map comprehension from another map.

**Comment by i4ki** [comment]
@geowatson yes, I fixed it here: https://github.com/mineiros-io/terramate/pull/867
Next release will have this fix and other improvements in the evaluator.
Thank you!

**Comment by i4ki** [comment]
Fixed by #867
